keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1338 @@
+				return nil, fmt.Errorf("valid platform matcher for image %v not found", img)
+			}
+		}
+	}
+	return imagesW, nil
+}
+
+</pre>
+
+
+### Windows platform matcher optional changes
+
+Windows platform matcher could optionally be extended to also check for host and UVM compatibility if hyperV isolation is being used based on the matrix described here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility 
+
+### Kubelet changes
+
+Kubelet needs to change how it talks about images from just the image name to (imageName, runtimeclass). As discussed in earlier sections, extending the Image struct in CRI to add a field for runtimeclass will make it easier to extend Kubelet API to include this information. 
+The following are scenarios where kubelet behavior needs to be extended to refer to image as its name and runtime class:
+
+1. Before running a container, kuebelet checks to see whether the required image for the container already exists or not by calling the **EnsureImageExists()** function.
+
+PullImageRequest has a field for ImageSpec which can hold info for the runtime class used to pull this image:
+
+<pre>
+message PullImageRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Authentication configuration for pulling the image.
+    AuthConfig auth = 2;
+    // Config of the PodSandbox, which is used to pull image in PodSandbox context.
+    PodSandboxConfig sandbox_config = 3;
+}
+</pre>
+
+2. Image status request/response can also make use of the newly added runtime class field in ImageSpec to specify the runtime class being used for the image. If none is specified, containerd will assume that the request is for the default runtime class defined for that platform in containerd.toml
+
+<pre>
+message ImageStatusRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Verbose indicates whether to return extra information about the image.
+    bool verbose = 2;
+}
+
+message ImageStatusResponse {
+    // Status of the image.
+    Image image = 1;
+    // Info is extra information of the Image. The key could be arbitrary string, and
+    // value should be in json format. The information could include anything useful
+    // for debug, e.g. image config for oci image based container runtime.
+    // It should only be returned non-empty when Verbose is true.
+    map<string, string> info = 2;
+}
+</pre>
+
+3. Image cleanup/kubelet garbage collection: Currently, kubelet GC will call into RemoveImage() on containerd side to remove unused images. kubelet will only pass the image name and containerd will currently delete all instances of the image. This needs to be extended on kubelet to ask for deleting (imageName, runtimeclass) to be more efficient. Optionally, we could add a behavior on containerd side to honor deletion of all versions of an image is that is was kubelet GC wants to do.
```

Comment:
> > This needs to be extended on kubelet to ask for deleting (imageName, runtimeclass) to be more efficient.
> 
> Is this possible?

I think this is GC behavior that would need to account for this once we change image to be referred by (imageName, runtimeclass) and I think this should be possible as runtimeclass field is added to the ImageSpec. If none is specified, containerd would try to delete the image with default runtime class.
cc @haircommander as he is working on altering kubelet GC behavior. I can work with you before changing this.

> 
> > Optionally, we could add a behavior on containerd side to honor deletion of all versions of an image is that is was kubelet GC wants to do.
> 
> I don't follow this sentence. What happens if all the images are cleaned up but some are still being used?

If some image is still being used, then containerd should not be successful in deleting the images, right? I think my sentence should be reworded as follows: 
"Optionally, we could add a behavior on containerd side to honor deletion of all versions of an image so that kubelet GC has a way to delete all versions of an image if it has identified that the images are not being used by any container/pod."

Does this answer the questions? Or am I missing smth?

Reply:
I am not super familiar with this area, but It seems like there could be some tricky edge cases here, if kubelet thinks it deleted an image but containerd doesn't and vice versa but I don't really know enough

Reply:
I think that problem already exists with kubelet and containerd behavior for image deletion - didn't mean to indicate any exisitng behavior would change. Reworded this to reduce confusion.